source/slang/slang-workspace-version.cpp

```diff
@@ -0,0 +1,358 @@
+    String currentPath;
+    String root;
+    void addSearchPath(String path)
+    {
+        while (path.getLength())
+        {
+            String canonicalPath;
+            Path::getCanonical(path, canonicalPath);
+            if (!paths.Add(canonicalPath))
+                break;
+            path = Path::getParentDirectory(path);
+            if (!path.startsWith(root))
+                break;
+        }
+    }
+};
+DocumentVersion* Workspace::openDoc(String path, String text)
+{
+    RefPtr<DocumentVersion> doc = new DocumentVersion();
+    doc->setText(text.getUnownedSlice());
+    doc->setURI(URI::fromLocalFilePath(path.getUnownedSlice()));
+    openedDocuments[path] = doc;
+    searchPaths.Add(Path::getParentDirectory(path));
+    invalidate();
+    return doc.Ptr();
+}
+
+void Workspace::init(List<URI> rootDirURI, slang::IGlobalSession* globalSession)
+{
+    for (auto uri : rootDirURI)
+    {
+        auto path = uri.getPath();
+        rootDirectories.add(path);
+        DirEnumerationContext context;
+        context.workList.add(path);
+        context.root = path;
+        auto fileSystem = Slang::OSFileSystem::getExtSingleton();
+        for (int i = 0; i < context.workList.getCount(); i++)
+        {
+            context.currentPath = context.workList[i];
+            fileSystem->enumeratePathContents(
+                context.currentPath.getBuffer(),
+                [](SlangPathType pathType, const char* name, void* userData)
+                {
+                    auto dirContext = (DirEnumerationContext*)userData;
+                    auto nameSlice = UnownedStringSlice(name);
+                    if (pathType == SLANG_PATH_TYPE_DIRECTORY)
+                    {
+                        dirContext->workList.add(Path::combine(dirContext->currentPath, name));
+                    }
+                    else if (nameSlice.endsWithCaseInsensitive(".slang") || nameSlice.endsWithCaseInsensitive(".hlsl"))
+                    {
+                        dirContext->addSearchPath(dirContext->currentPath);
+                    }
+                },
+                &context);
+        }
+        searchPaths = _Move(context.paths);
+    }
+    slangGlobalSession = globalSession;
+}
+
+void Workspace::invalidate() { currentVersion = nullptr; }
+
+int parseInt(UnownedStringSlice text, Index& pos)
+{
+    int result = 0;
+    while (text[pos] == ' ' && pos < text.getLength())
+    {
+        pos++;
+        continue;
+    }
+    while (pos < text.getLength())
+    {
+        if (text[pos] >= '0' && text[pos] <= '9')
+        {
+            result *= 10;
+            result += text[pos] - '0';
+            pos++;
+        }
+        else
+        {
+            break;
+        }
+    }
+    return result;
+}
+
+void parseDiagnostics(Dictionary<String, DocumentDiagnostics>& diagnostics, String compilerOutput)
+{
+    List<UnownedStringSlice> lines;
+    StringUtil::split(compilerOutput.getUnownedSlice(), '\n', lines);
```

Comment:
Might be better to use StringUtil::calcLines as takes into account line ending styles.